<?xml version="1.0"?>

<robot name="diffdrive" xmlns:xacro="http://www.ros.org/wiki/xacro">

  <!-- include gazebo plugins -->
  <xacro:include filename="$(find homework5)/urdf/diffdrive.gazebo" />

  <xacro:property name="base_width" value="0.15" />
  <xacro:property name="base_length" value="0.2" />
  <xacro:property name="base_height" value="0.05" />
  <xacro:property name="wheel_radius" value="0.05" />
  <xacro:property name="wheel_width" value="0.02" />
  <xacro:property name="wheel_gap" value="0.0" />
  <xacro:property name="wheel_setback" value="0.05" />
  <xacro:property name="lidar_width" value="0.02" />
  <xacro:property name="lidar_length" value="0.02" />
  <xacro:property name="lidar_height" value="0.02" />
  <xacro:property name="lidar_setback" value="0.0" />


  <!--~~~~~~~~~~~~~~~~~~~~~~~~ ~ ROBOT BASE ~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~-->

  <!-- Make a box for robot base -->

  <gazebo reference="base_link">
    <material>Gazebo/White</material>
  </gazebo>
  <link name="base_link">

  	<!-- Draw imaginary box around the robot for collision detection -->
    <collision>
      <origin xyz="0 0 0.0" rpy="0 0 0"/>
      <geometry>
       <box size="${base_length} ${base_width} ${base_height}"/>
      </geometry>
    </collision>

    <!-- Describe how to draw the box in 3d -->
    <visual>
      <origin xyz="0 0 0.0" rpy="0 0 0"/>
      <geometry>
	       <box size="${base_length} ${base_width} ${base_height}"/>
      </geometry>

    </visual>

    <collision name='caster_collision'>
      <origin xyz="${-base_length/2 + wheel_radius/4} 0 ${-wheel_radius/2}" rpy="0 0 0"/>
      <geometry>
          <sphere radius="${wheel_radius/2}"/>
      </geometry>

      <surface>
        <friction>
          <ode>
            <mu>0</mu>
            <mu2>0</mu2>
            <slip1>1.0</slip1>
            <slip2>1.0</slip2>
          </ode>
        </friction>
      </surface>
    </collision>

    <visual name='caster_visual'>
      <origin xyz="${-base_length/2 + wheel_radius/4} 0 ${-wheel_radius/2}" rpy="0 0 0"/>
      <geometry>
        <sphere radius="${wheel_radius/2}"/>
      </geometry>
    </visual>

    <inertial>
      <mass value="0.1" />  <!-- kilograms -->
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="0.00354" ixy="0" ixz="0" iyy="0.005208" iyz="0" izz="0.00208"/>
    </inertial>

  </link>

  <link name="left_link">
      <visual>
        <geometry>
          <cylinder length="${wheel_width}" radius="${wheel_radius}"/>
        </geometry>
        <origin rpy="${pi/2} 0 0" xyz="0 0 0"/>
      </visual>

      <collision>
        <geometry>
          <cylinder length="${wheel_width}" radius="${wheel_radius}"/>
        </geometry>
        <origin rpy="${pi/2} 0 0" xyz="0 0 0"/>


      </collision>

      <inertial>
        <mass value=".2" />
        <origin rpy="${pi/2} 0 0" xyz="0 0 0"/>
        <inertia ixx="0.00032916" ixy="0" ixz="0" iyy="0.00032916" iyz="0" izz="0.000625"/>
      </inertial>
    </link>

    <link name="right_link">
        <visual>
          <geometry>
            <cylinder length="${wheel_width}" radius="${wheel_radius}"/>
          </geometry>
          <origin rpy="${pi/2} 0 0" xyz="0 0 0"/>
        </visual>

        <collision>
          <geometry>
            <cylinder length="${wheel_width}" radius="${wheel_radius}"/>
          </geometry>
          <origin rpy="${pi/2} 0 0" xyz="0 0 0"/>

        </collision>

        <inertial>
          <mass value=".2" />
          <origin rpy="${pi/2} 0 0" xyz="0 0 0"/>
          <inertia ixx="0.00032916" ixy="0" ixz="0" iyy="0.00032916" iyz="0" izz="0.000625"/>
        </inertial>
      </link>



  <!--~~~~~~~~~~~~~~~~~~~~~~~ ~ Lidar ~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~-->

  <gazebo reference="laser_link">
    <material>Gazebo/Orange</material>
  </gazebo>
  <link name="laser_link">

    <!--<collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
       <box size="${lidar_length} ${lidar_width} ${lidar_height}"/>
      </geometry>
    </collision>-->

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
	       <box size="${lidar_length} ${lidar_width} ${lidar_height}"/>
      </geometry>
      <material name="red"/>
    </visual>

    <inertial>
      <mass value="1e-6" />
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" />
    </inertial>
  </link>

  <joint name="joint1" type="fixed">      <!-- Fused connection? (using continuous tilts view)-->
    <parent link="base_link"/>            <!--      robot link = parent     -->
    <child link="laser_link"/>     <!-- forward camera link = child  -->
    <origin xyz="${base_length/2 - lidar_length/2 - lidar_setback} 0 ${lidar_height/2 + base_height/2}" rpy="0 0 0"/>   <!-- where child is relative to parent -->
    <axis xyz="0 0 0"/>                   <!-- ...describes the axis of rotation? -->
  </joint>

  <joint name="left_joint" type="continuous">
      <parent link="base_link"/>
      <child link="left_link"/>
      <origin rpy="0 0 0" xyz="${base_length/2 - wheel_setback} ${base_width/2 + wheel_width/2 + wheel_gap} 0.0"/>
      <axis xyz="0 1 0"/>
      <dynamics damping="0" friction="0" spring_stiffness="1" />
  </joint>


  <joint name="right_joint" type="continuous">
      <parent link="base_link"/>
      <child link="right_link"/>
      <origin rpy="0 0 0" xyz="${base_length/2 - wheel_setback} ${-base_width/2 - wheel_width/2 - wheel_gap} 0.0"/>
      <axis xyz="0 1 0"/>
      <dynamics damping="0" friction="0" spring_stiffness="0" />
  </joint>

</robot>
